### PR TITLE
Prevent action conflicts for exec-starlark-exec transitions

### LINKED ISSUE #13464: Conflicting actions for `cc_proto_library` with transition in Starlark rule dependency
### Description of the problem / feature request:

As of 7acf9ea3fea4850d21114d84fbc5b663d105d3c7, Bazel generates conflicting actions when a custom Starlark rule depends on a `cc_proto_library` target in the `exec` configuration if there is a transition applied to the library.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

```bash
$ cat WORKSPACE
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")

http_archive(
    name = "com_google_protobuf",
    sha256 = "7892a35d979304a404400a101c46ce90e85ec9e2a766a86041bb361f626247f5",
    strip_prefix = "protobuf-3.16.0",
    url = "https://github.com/protocolbuffers/protobuf/archive/v3.16.0.tar.gz",
)

load("@com_google_protobuf//:protobuf_deps.bzl", "protobuf_deps")

protobuf_deps()

$ cat BUILD.bazel
load(":rules.bzl", "exec_with_library_dep", "apply_transition")

proto_library(
    name = "proto",
    srcs = ["empty.proto"],
    deps = [],
)

cc_proto_library(
    name = "cc_proto",
    deps = [":proto"],
)

apply_transition(
    name = "cc_proto_with_transition",
    native_target = ":cc_proto",
)

exec_with_library_dep(
    name = "error",
    out = "dummy.gen",
    library = ":cc_proto_with_transition",
)

$ touch empty.proto

$ cat rules.bzl
_COPT_BUILD_SETTING = "//command_line_option:copt"

def _set_foobar_copt(settings, attr):
    return {_COPT_BUILD_SETTING: ["-DFOOBAR"]}

set_foobar_copt = transition(
    implementation = _set_foobar_copt,
    inputs = [],
    outputs = [_COPT_BUILD_SETTING],
)

def _apply_transition_impl(ctx):
    pass

apply_transition = rule(
        implementation = _apply_transition_impl,
        attrs = {
            "native_target": attr.label(
                cfg = set_foobar_copt,
            ),
            "_allowlist_function_transition": attr.label(
                default = "@bazel_tools//tools/whitelists/function_transition_whitelist",
            ),
        },
    )


def _exec_with_library_dep_impl(ctx):
    ctx.actions.write(ctx.outputs.out, "")

exec_with_library_dep = rule(
    _exec_with_library_dep_impl,
    attrs = {
        "library": attr.label(
            mandatory = True,
            cfg = "exec",
        ),
        "out": attr.output(
            mandatory = True,
        ),
    },
)

$ bazel build //:error
```

You can also clone https://github.com/fmeum/cc_proto_library_transition_issue.

The command will generate the following error:

```
ERROR: file 'external/com_google_protobuf/_objs/protobuf_lite/int128.o' is generated by these conflicting actions:
Label: @com_google_protobuf//:protobuf_lite
RuleClass: cc_library rule
Configuration: 8d1a3355bd0201255e646e7ddf342ec528414ebbf2f1feab6e6b20aced50487e, 64b9564990ccaa73f2844f556c40131bd590d44ee9288aea2e3f3beb0c1a956b
Mnemonic: CppCompile
Action key: 4a432e1690a4da20c2fe249ae6554cf201d9792e4375c5a945c9962a8f2ec395, 9dc03c391082ba0e9a3e692ac5cc7257e5abab2a8de3eb5226db9bc4a4ba9f50
Progress message: Compiling src/google/protobuf/stubs/int128.cc
PrimaryInput: File:[/home/fhenneke/.cache/bazel/_bazel_fhenneke/3665225be4dbfed855cbc8b2bd45f83a/external/com_google_protobuf[source]]src/google/protobuf/stubs/int128.cc
PrimaryOutput: File:[[<execution_root>]bazel-out/k8-opt-exec-2B5CBBC6-ST-cbcaac683f1b/bin]external/com_google_protobuf/_objs/protobuf_lite/int128.o
Owner information: ConfiguredTargetKey{label=@com_google_protobuf//:protobuf_lite, config=BuildConfigurationValue.Key[8d1a3355bd0201255e646e7ddf342ec528414ebbf2f1feab6e6b20aced50487e]}, ConfiguredTargetKey{label=@com_google_protobuf//:protobuf_lite, config=BuildConfigurationValue.Key[64b9564990ccaa73f2844f556c40131bd590d44ee9288aea2e3f3beb0c1a956b]}
MandatoryInputs: are equal
Outputs: are equal
ERROR: com.google.devtools.build.lib.skyframe.ArtifactConflictFinder$ConflictException: com.google.devtools.build.lib.actions.MutableActionGraph$ActionConflictException: for external/com_google_protobuf/_objs/protobuf_lite/int128.o, previous action: action 'Compiling src/google/protobuf/stubs/int128.cc', attempted action: action 'Compiling src/google/protobuf/stubs/int128.cc'
INFO: Elapsed time: 0.121s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded, 0 targets configured)
```

### What operating system are you running Bazel on?

Ubuntu 20.10

### What's the output of `bazel info release`?

The issue reproduces with `release 4.0.0` as well as master @ 7caa01fcb69b1b3d5c67476b100b13e82097227d, but not with `release 3.7.2`. 

Using `bazel-bisect.sh`, I have identified the breaking commit as 7acf9ea3fea4850d21114d84fbc5b663d105d3c7, which enabled exec transitions on proto rules.

### PR DESCRIPTION
**Pull request by fmeum**
With this commit, an execution transition will recompute the transition
hash appended to the output directory name if set by any previous
Starlark transition. This also takes into account that the Starlark
transition may transitively affect an option --foo_bar after the exec
transition if it previously affected the corresponding host option
--host_foo_bar.

Previously, an execution transition would reuse the existing hash, which
no longer reflected the current state of native options correctly and
thus led to action conflicts.

Fixes #13464.

### CODE REVIEW

File: AUTHORS [hunk 1]
```diff
@@ -20,6 +20,7 @@ liuhuahang <liuhuahang@zerus.co>
 David Mankin <dmankin@atlassian.com>
 Kamal Marhubi <kamal@marhubi.com>
 Carl Mastrangelo <carl-mastrangelo@users.noreply.github.com>
+Fabian Meumertzheim <fabian@meumertzhe.im>
```
Comment (aiuto) [inline AUTHORS:23]: Please do not add yourself to AUTHORS.
By agreeing to the CLA you relinquish copyright claim.

File: CONTRIBUTORS [hunk 2]
```diff
@@ -61,6 +61,7 @@ Carl Mastrangelo <carl-mastrangelo@users.noreply.github.com>
 Michajlo Matijkiw <michajlo@google.com>
 Iain McGinniss <iainmcgin@google.com>
 Julio Merino <jmmv@google.com>
+Fabian Meumertzheim <fabian@meumertzhe.im>
```
Comment (aiuto) [inline CONTRIBUTORS:64]: new names should go at the end.

File: AUTHORS (hunk 1 above)
Comment (fmeum) [reply AUTHORS]: Done

File: CONTRIBUTORS (hunk 2 above)
Comment (fmeum) [reply CONTRIBUTORS]: Done

**Comment by katre** [comment]
@sdtwigg is reviewing this.

**Comment by gregestren** [comment]
Continuing the conversation at https://github.com/bazelbuild/bazel/issues/13464 to clarify approaches.

**Comment by sdtwigg** [comment]
To clarify, I/we have not forgotten about this. There are just a lot of potential concerns (mostly with where to take this API in the future) that we want to address before proceeding.

**Comment by fmeum** [comment]
@sdtwigg Do you think that we'll be able to get some version of this (possibly without the `host_*` accounting) into 5.0?

**Comment by fmeum** [comment]
@gregestren I'm seeing unexpected NullPointerExceptions in the tests: https://storage.googleapis.com/bazel-untrusted-buildkite-artifacts/36b11777-1957-4dc8-9914-3bdd82aa6e4a/src%5Ctest%5Cjava%5Ccom%5Cgoogle%5Cdevtools%5Cbuild%5Clib%5Crules%5Ccpp%5CCcLibraryConfiguredTargetTest%5Cattempt_1.log.
Do you have an idea why they arise?

**Comment by gregestren** [comment]
I'm not sure. Which line is `FunctionTransitionUtil.java:442` in your dev build?

**Comment by fmeum** [comment]
> I'm not sure. Which line is `FunctionTransitionUtil.java:442` in your dev build?

This one: https://github.com/bazelbuild/bazel/blob/01cbfce9ffc9bcc50ca7d6e397305a52a29147aa/src/main/java/com/google/devtools/build/lib/analysis/starlark/FunctionTransitionUtil.java#L442

**Comment by gregestren** [comment]
I'm not sure why the tests get `NullPointerException`s, I'm sorry. Maybe some mock test infrastructure is missing an important field?

More broadly, can we pause this while we work out https://github.com/bazelbuild/bazel/issues/13464#issuecomment-929671793 and the discussion at https://github.com/bazelbuild/bazel/pull/13587#issuecomment-933954517?

The right next steps, in my mind, are for @sdtwigg and I to seed a proper design doc for all of us to focus our thoughts on. I want to be careful not to eliminate action conflicts by an endless layer of ad hoc changes.

**Comment by fmeum** [comment]
> I'm not sure why the tests get `NullPointerException`s, I'm sorry. Maybe some mock test infrastructure is missing an important field?
> 
> More broadly, can we pause this while we work out https://github.com/bazelbuild/bazel/issues/13464#issuecomment-929671793 and the discussion at https://github.com/bazelbuild/bazel/pull/13587#issuecomment-933954517?
> 
> The right next steps, in my mind, are for @sdtwigg and I to seed a proper design doc for all of us to focus our thoughts on. I want to be careful not to eliminate action conflicts by an endless layer of ad hoc changes.

Sounds very good to me, the proposal seems very reasonable and would resolve a cluster of usability and performance issues around transitions, beyond action conflicts.

Let me know when you have started a draft of the design doc and I will see what I can contribute.

**Comment by sdtwigg** [comment]
Ok, I was originally confused by the part of the this change (and corresponding discussion) regarding transitioning on `--host_*` values. (For reference, that seems tricky but we should discuss the merits as a separate issue.)

I need to find a proper issue to post the longer-form summary of our plan around ST-hash computation; however, for this specific sub-issue, I was planning to fix this issue by moving transitionDirectoryNameFragment out of BuildOptions and into BuildConfigurationValue (previously called BuildConfiguration).

`transitionDirectoryNameFragment` should wholly be a function of BuildOptions and it is fragile to require all transitions to remember to update it when it touches associated options, native or otherwise. BuildConfigurationValue is already a container for potentially-expensive-to-compute values based off BuildOptions so movement there is a natural fit (e.g. notably OutputDirectories, which ultimately consumed transitionDirectoryNameFragment is already part of BuildConfigurationValue!)

This essentially makes this PR irrelevant (although considering it did expose other potential performance problems with the the execution transition).

**Comment by sdtwigg** [comment]
See https://github.com/bazelbuild/bazel/issues/14023#issuecomment-956905920

**Comment by fmeum** [comment]
> Ok, I was originally confused by the part of the this change (and corresponding discussion) regarding transitioning on `--host_*` values. (For reference, that seems tricky but we should discuss the merits as a separate issue.)
> 
> I need to find a proper issue to post the longer-form summary of our plan around ST-hash computation; however, for this specific sub-issue, I was planning to fix this issue by moving transitionDirectoryNameFragment out of BuildOptions and into BuildConfigurationValue (previously called BuildConfiguration).
> 
> `transitionDirectoryNameFragment` should wholly be a function of BuildOptions and it is fragile to require all transitions to remember to update it when it touches associated options, native or otherwise. BuildConfigurationValue is already a container for potentially-expensive-to-compute values based off BuildOptions so movement there is a natural fit (e.g. notably OutputDirectories, which ultimately consumed transitionDirectoryNameFragment is already part of BuildConfigurationValue!)
> 
> This essentially makes this PR irrelevant (although considering it did expose other potential performance problems with the the execution transition).

I fully agree that moving the `transitionDirectoryNameFragment` computation is a much better solution. I will close this PR in favor of this approach. Depending on how it is implemented, the `host_*` options might lead to action conflicts again, but we can just keep that in mind and e.g. reuse the test case from this PR.

**Comment by fmeum** [comment]
@sdtwigg Should I create a new PR containing just the test? It could serve as a useful regression test for upcoming changes to transition logic.

**Pull request closed without merging**